sdks/java/fn-execution/src/main/java/org/apache/beam/sdk/fn/stream/DataStreams.java

```diff
@@ -186,14 +186,16 @@ public DataStreamDecoder(Coder<T> coder, PrefetchableIterator<ByteString> inputS
      * ByteString} in the underlying {@link ByteString} {@link Iterator iterator} and decoding
      * elements till at the next boundary.
      */
-    public List<T> decodeFromChunkBoundaryToChunkBoundary() {
+    public WeightedList<T> decodeFromChunkBoundaryToChunkBoundary() {
       inbound.currentStream = inputByteStrings.next().newInput();
       inbound.position = 0;
       try {
         InputStream previousStream = inbound.currentStream;
-        List<T> rvals = new ArrayList<>();
+        WeightedList<T> rvals = new WeightedList<>(new ArrayList<>(), 0L);
         while (previousStream == inbound.currentStream && inbound.currentStream.available() != 0) {
-          rvals.add(next());
+          inbound.resetCounter();
```

Comment:
Nit: rather than having a stateful reset, you could record the bytes before and after and take the difference.

Reply:
Done, thanks!